model/generate.go

```diff
@@ -275,25 +270,171 @@ func (g *GeneratedProject) saveNewBuildsAndTasks(ctx context.Context, v *Version
+	}
+
+	simulatedGraph, err := g.simulateNewTasks(existingTasksGraph, v, p, projectRef)
+	if err != nil {
+		return errors.Wrap(err, "simulating new tasks")
+	}
+
+	if cycles := simulatedGraph.Cycles(); len(cycles) > 0 {
+		return errors.Wrapf(DependencyCycleError, "'%s'", cycles)
+	}
+
+	return nil
+}
+
+// simulateNewTasks adds the tasks we're planning to add to the version to the graph and
+// adds simulated edges from each task that depends on the generator to each of the generated tasks.
+func (g *GeneratedProject) simulateNewTasks(graph task.DependencyGraph, v *Version, p *Project, projectRef *ProjectRef) (task.DependencyGraph, error) {
+	newTasks := g.getNewTasksWithDependencies(v, p)
+
+	taskIDs, err := getTaskIdTables(v, p, newTasks, projectRef.Identifier)
+	if err != nil {
+		return graph, errors.Wrap(err, "getting task ids")
+	}
+
+	graph = addTasksToGraph(newTasks.ExecTasks, graph, p, taskIDs)
+	return g.addDependencyEdgesToGraph(newTasks.ExecTasks, v, p, graph, taskIDs)
+}
+
+// getNewTasksWithDependencies returns the generated tasks and their recursive dependencies.
+func (g *GeneratedProject) getNewTasksWithDependencies(v *Version, p *Project) TaskVariantPairs {
+	newTVPairs := TaskVariantPairs{}
+	for _, bv := range g.BuildVariants {
+		newTVPairs = appendTasks(newTVPairs, bv, p)
+	}
+
+	var err error
+	newTVPairs.ExecTasks, err = IncludeDependencies(p, newTVPairs.ExecTasks, v.Requester)
+	grip.Warning(message.WrapError(err, message.Fields{
+		"message": "error including dependencies for generator",
+		"task":    g.Task.Id,
+	}))
+
+	return newTVPairs
+}
+
+// addTasksToGraph adds tasks to the graph and adds dependency edges from each task task to each of its dependencies.
+func addTasksToGraph(tasks TVPairSet, graph task.DependencyGraph, p *Project, taskIDs TaskIdConfig) task.DependencyGraph {
+	for _, newTask := range tasks {
+		graph.AddTaskNode(task.TaskNode{
+			ID:      taskIDs.ExecutionTasks.GetId(newTask.Variant, newTask.TaskName),
+			Name:    newTask.TaskName,
+			Variant: newTask.Variant,
+		})
+	}
+
+	bvts := make([]BuildVariantTaskUnit, 0, len(tasks))
+	for _, node := range graph.Nodes() {
+		bvt := p.FindTaskForVariant(node.Name, node.Variant)
+		if bvt != nil {
+			bvts = append(bvts, *bvt)
+		}
+	}
+
+	for _, dep := range dependenciesForTaskUnit(bvts) {
+		dep.From.ID = taskIDs.ExecutionTasks.GetId(dep.From.Variant, dep.From.Name)
+		dep.To.ID = taskIDs.ExecutionTasks.GetId(dep.To.Variant, dep.To.Name)
+		graph.AddEdge(dep.From, dep.To, dep.Status)
+	}
+
+	return graph
+}
+
+// addDependencyEdgesToGraph adds edges from the tasks that depend on the generator to activated generated tasks.
+func (g *GeneratedProject) addDependencyEdgesToGraph(newTasks TVPairSet, v *Version, p *Project, graph task.DependencyGraph, taskIDs TaskIdConfig) (task.DependencyGraph, error) {
+	activatedNewTasks, err := g.filterInactiveTasks(newTasks, v, p)
```

Comment:
There are two types of dependency cycles that could be created by generated tasks. One is if the new project contains explicit cycles. That will already be caught by the project validator. 
The second type is when we add dependencies for the tasks that depend on the generator on the generated tasks. This isn't part of the project (I tried adding it a few times but failed 😞) so we have to do this simulation thing. Since #4793 we only do this for tasks that are being activated _at generate time_, so the validation just has to simulate these new dependencies for tasks being activated now.